src/DocumentTemplate/ustr.py

```diff
@@ -12,7 +12,7 @@
 ##############################################################################
 """ustr function."""
 
-nasty_exception_str = getattr(Exception.__str__, 'im_func', None)
+nasty_exception_str = getattr(Exception.__str__, '__func__', None)
```

Comment:
Interestingly, this is only non-None on PyPy2. Everywhere else, it will be `None`.
```console
$ python2.7 -c 'Exception.__str__.__func__'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
AttributeError: 'wrapper_descriptor' object has no attribute '__func__'
$ python3.5 -c 'Exception.__str__.__func__'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
AttributeError: 'wrapper_descriptor' object has no attribute '__func__'
$ python3.6 -c 'Exception.__str__.__func__'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
AttributeError: 'wrapper_descriptor' object has no attribute '__func__'
$ python3.9 -c 'Exception.__str__.__func__'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
AttributeError: 'wrapper_descriptor' object has no attribute '__func__'
$ pypy -c 'Exception.__str__.__func__'
$ pypy3 -c 'Exception.__str__.__func__'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
AttributeError: 'function' object has no attribute '__func__'
```

This perhaps makes the updated `if` test below somewhat confusing to follow...

Reply:
You are right!

The apparent purpose of the code is to recognize exceptions and render them specially (apparently because their `__str__` returned a wrong type in some cases). This can be done much clearer and safer. I will propose a different solution - completely avoiding `im_func/__func__`.